include/picongpu/plugins/radiation/Radiation.kernel

```diff
@@ -167,53 +168,39 @@ namespace picongpu
-                            /* since a race condition can occur if "continue loop" is called,
-                             *  all threads must wait for the selection of a new frame
-                             *  until all threads have evaluated "isValid"
-                             */
-                            worker.sync();
-
-                            auto onlyMaster = lockstep::makeMaster(worker);
-
-                            /* The Master process (thread 0) in every thread block is in
-                             * charge of loading a frame from
-                             * the current super cell and evaluate the total number of
-                             * particles in this frame.
-                             */
-                            onlyMaster([&]() { counter_s = 0; });
-
-                            worker.sync();
-
-                            // loop over all particles in the frame
-                            auto forEachParticle = lockstep::makeForEach<frameSize>(worker);
-
-                            forEachParticle(
-                                [&](uint32_t const linearIdx)
-                                {
-                                    // only threads with particles are running
-                                    if(linearIdx < particlesInFrame)
+                            globalOffset + ((superCellIdx - guardingSuperCells) * SuperCellSize::toRT()));
+
+                        namespace particleAccAlgos = pmacc::particles::algorithm::acc;
+                        auto forEachFrameInSupercell
+                            = particleAccAlgos::makeForEachFrame<particleAccAlgos::Forward>(worker, pb, superCellIdx);
+
+                        if(!forEachFrameInSupercell.hasParticles())
+                            continue;
+
+                        forEachFrameInSupercell(
+                            /* iterate over the frame list of the current supercell */
+                            [&](auto const& lockstepWorker, auto& frameCtx)
+                            {
+                                lockstep::makeMaster(worker)([&]() { counter_s = 0; });
+
+                                // wait for particle counter update
+                                worker.sync();
+
+                                // loop over all particles in the frame
+                                auto forEachParticleInFrame = forEachFrameInSupercell.lockstepForEach();
+
+                                forEachParticleInFrame(
+                                    [&](lockstep::Idx const linearIdx)
                                     {
-                                        auto par = frame[linearIdx];
+                                        /* modulo is required if the kernel is started with more workers than particle
+                                         * in a supercell
+                                         */
+                                        auto par = frameCtx[linearIdx][linearIdx % frameSize];
```

Comment:
@psychocoderHPC what does `[linearIdx % frameSize]` do/access?

Reply:
The for each particle functor has the possibility to use a skip list if you execute the algorithm with more threads than particles in a frame. In this case we select a particle in the frame. 
offline discussed: the radiation shared memory structure is not ready to use this feature but to avoid later problems because we access a particle in the frame outside of the range the modulo is already included.